Add ed25519 support

johannww opened the pull request:
#### Type of change

- New feature

#### Description

Due to some considering NIST curves insecure and
Golang having ed25519 native support, there was
not a reason for not implementing in Fabric.

Tests cases for ed25519 were also added. Since
ed25519 key derivation is not called by any
function, I left as a TODO.

As I am working on ed25519 support for node fabric-gateway,
I needed to add ed25519 support for cryptogen also,
aiming to pass tests with certificates containing
ed25519 keys. Since the node fabric-gateway tests
generate their crypto material with cryptogen, I
adapted cryptogen to support ed25519 keys.

#### Additional details

Some tests concerning ed25519 key generation and certificate parsing
were added.

#### Related issues

<!--- Include a link to any associated issues, e.g. Jira issue or approved rfc. -->

<!---
#### Release Note
If change impacts current users, uncomment Release Note heading and provide
release note text.
Also, copy release note text into the release specific /release_notes file.
-->

johannww committed:
Add ed25519 support

Due to some considering NIST curves insecure and
Golang having ed25519 native support, there was
not a reason for not implementing in Fabric.

Tests cases for ed25519 were also added. Since
ed25519 key derivation is not called by any
function, I left as a TODO.

As I am working on ed25519 support for node fabric-gateway,
I needed to add ed25519 support for cryptogen also,
aiming to pass tests with certificates containing
ed25519 keys. Since the node fabric-gateway tests
generate their crypto material with cryptogen, I
adapted cryptogen to support ed25519 keys.

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

yacovm commented:
I generally support this idea, and am in favor of it, but unfortunately adding support for ed25519 in a safe manner is more complex than just adding the code for generating signatures and verifying them. 

If you could atomically and instantly upgrade all nodes of a Fabric network to support ed25519 then there wouldn't have been any problem at all. 

But Fabric nodes may be upgraded at different times, and some nodes might be offline or unreachable and then afterwards be reachable and online again and these nodes will receive the ed25519 signatures and they will reject them.
On the other hand, the nodes that were upgraded, will not reject them, and you will have a fork in the blockchain.

The key problem is that we need a way for Fabric validation to understand that "until a certain point in the blockchain, all ed25519 signatures are invalid" and "starting from a certain point in the blockchain, all ed25519 signatures are valid".

In Fabric, this is done using the [capabilities](https://hyperledger-fabric.readthedocs.io/en/latest/capabilities_concept.html) feature.  Essentially, you "mark" in the channel config a certain configuration property and then the nodes know when to use that property and when not to use it.

yacovm commented:
Aside from my comment above, the PR is missing [integration tests](https://github.com/hyperledger/fabric/tree/main/integration) that generate ed25519 crypto material, and run a network and transact to ensure everything is good.

yacovm commented on internal/cryptogen/msp/msp_test.go:
```diff
@@ -110,10 +112,10 @@ func testGenerateLocalMSP(t *testing.T, nodeOUs bool) {
 
 	tlsCA.Name = "test/fail"
-	err = msp.GenerateLocalMSP(testDir, testName, nil, signCA, tlsCA, msp.CLIENT, nodeOUs)
+	err = msp.GenerateLocalMSP(testDir, testName, nil, signCA, tlsCA, msp.CLIENT, nodeOUs, ECDSA)
```
I see that you didn't changed the MSP (top level `msp package`).

The MSP needs to be versioned like [here](https://github.com/hyperledger/fabric/blob/main/msp/mspimpl.go#L112-L139) from the aforementioned reasons.

johannww replied:
Thanks for all your comments! I intend to add more work to it during the following weeks.

ale-linux commented:
Incidentally, have you checked that the verification takes care of 

https://datatracker.ietf.org/doc/html/rfc8032#section-8.4

We had to do this manually for ecdsa

johannww commented:
> Incidentally, have you checked that the verification takes care of
> 
> https://datatracker.ietf.org/doc/html/rfc8032#section-8.4
> 
> We had to do this manually for ecdsa

It seems that golang has already implemented it:

https://github.com/golang/go/blob/4e79f06dac712d35d67d72777dae6df6d8c97888/src/crypto/ed25519/ed25519.go#L220

https://github.com/golang/go/blob/4e79f06dac712d35d67d72777dae6df6d8c97888/src/crypto/ed25519/ed25519_test.go#L165

In previous versions they even had a comment about preventing signature malleability:

https://github.com/golang/go/blob/go1.15/src/crypto/ed25519/ed25519.go#L235

bestbeforetoday commented:
@johannww On ed25519 support for node fabric-gateway... I'm sure you already know the signing implementation is entirely pluggable. The client application can supply its own (ed25519) signing implementation and the API doesn't need to have an implementation built in. However, I would be quite happy to include this in the API. Just note that there need to be implementations in all three client languages (Go, Type/JavaScript and Java).

johannww committed:
Add initial integration tests for ed25519

Based on the raft integration test, they are
mainly the same tests, but executed in a network created with
ed25519 crypto material.

Note that in this commit we did not added the ed25519 integration
test network creation, implemented in the "integration/nwo" folder.
The present commit only add the tests.

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

johannww committed:
Add ed25519 channel and msp capabilities

This commit has the changes in the channel capabilities,
introducing channel capabilities version V2_0,
which invokes a new msp version "MSPv_1_4_4". This is
the only msp version which accepts ed25519 identities.

This ensures that, during the upgrade process to support
ed25519, the blockchain will not become inconsistent.

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

johannww committed:
Enhance capabilities and add ed25519 integration tests

This commit will be forcibly squashed afterwards, but it is
intended to save partially capabilities enhancements and
ed25519 integration tests.

About Channel capabilities, I renamed the Channel capabilities that
support ed25519 from v2_1 to v2_4. Likewise, the MSP version was also
renamed to MSPv2_4.

Integration tests and the needed modifications were also added,
to ensure proper behavior of the Channel v2_4 capabilities.
The test's main goal was to prove that a peer without ed25519
support will stop validating blocks for the a channel with
v2_4 capabilities.

Cryptogen flag "--keyalg" was removed from my previous commit
to keep it simple. I also added a field on the parsed crypto-
config.yaml to generate ed25519 keys.

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

yacovm commented on msp/factory.go:
```diff
@@ -60,8 +60,8 @@ func New(opts NewOpts, cryptoProvider bccsp.BCCSP) (MSP, error) {
 			return newBccspMsp(MSPv1_4_3, cryptoProvider)
-		case MSPv1_4_4:
-			return newBccspMsp(MSPv1_4_4, cryptoProvider)
+		case MSPv2_4:
```
We cannot put this in 2.4, since 2.4 already exists. 

I suggest we put this in Fabric v3.0.

yacovm commented on msp/mspimplsetup.go:
```diff
@@ -656,6 +659,22 @@ func (msp *bccspmsp) setupV142(conf *m.FabricMSPConfig) error {
 	return nil
 }
 
+func (msp *bccspmsp) setupV24(conf *m.FabricMSPConfig) error {
```
Lets put setupV3

yacovm commented on integration/nwo/standard_networks.go:
```diff
@@ -181,6 +181,30 @@ func MultiChannelBasicSolo() *Config {
 	return config
 }
 
+func Ed25519Solo() *Config {
```
why do you need a config for ED25519 ? There is nothing unique about this config that is related to ed25519. In the integration test you can just take an existing config, no?

johannww replied:
I probably don't. I only created there as a temporary solution. Peers can be added externally to this configuration function anyway.

yacovm commented on integration/nwo/network.go:
```diff
@@ -1439,6 +1461,79 @@ func (n *Network) peerCommand(command Command, tlsDir string, env ...string) *ex
 	return cmd
 }
 
+func (n *Network) downloadOldBinaries(fabricVersion string) error {
```
why do you need this `downloadOldBinaries` ?

yacovm commented on integration/nwo/network.go:
```diff
@@ -1439,6 +1461,79 @@ func (n *Network) peerCommand(command Command, tlsDir string, env ...string) *ex
+	return nil
+}
+
+func (n *Network) oldPeerCommand(command Command, fabricVersion string, tlsDir string, env ...string) *exec.Cmd {
```
why do you need to have this `oldPeerCommand` command?

johannww replied:
In the integration test integration/lifecycle/capabilities_endorse_test.go I upgrade the channel capabilities and intend to show that a peer that does not have source code to support it won't endorse any more transactions. Therefore, one of the peers runs with a fabric v2.4.5 binary, downloaded from Github.

yacovm replied:
did you mean to say "does not" instead of "does" ? I still don't understand, why would the peer not endorse transactions? If you enable a capability and a peer cannot support it, it simply crashes.

johannww replied:
I assumed that the peer crash should be proven. If you consider this verification unnecessary I can remove it.

yacovm replied:
yes please remove

yacovm commented on integration/ed25519/config_test.go:
```diff
@@ -0,0 +1,1859 @@
+/*
+Copyright IBM Corp All Rights Reserved.
```
I think a **single** integration test that:

- Has peers and orderers in it
- Transacts on a channel
- The channel contains both ECDSA and ed25519 identities and they're both used in the endorsement policy

Is sufficient.

johannww replied:
That is my intention. This commit was not meant as a final solution yet but a work in progress. The folder integration/ed25519 will be removed.

johannww replied:
By "peers and oderer**S**" do you mean I should use a raft configuration with at least two orderers?

yacovm replied:
Knowing the Raft code, you don't have to use several Raft nodes to test ed25519. One is sufficient. But you can use three, I think it's easier from a copy-paste aspect that to use one.

yacovm commented on integration/ed25519/client.go:
```diff
@@ -0,0 +1,77 @@
+	return blk
+}
+
+func CreateBroadcastEnvelope(n *nwo.Network, entity interface{}, channel string, data []byte) *common.Envelope {
```
we have these functions already in the code, I don't think we need to add them again in another place. Let's just use the functions that exist

yacovm commented:
Ok i understand this is still work in progress. So please tag me once you feel this is ready. 
Take into account that it will take quite a while for this to be merged, and also even more time until this will be included into any release, because if it will go to any release it will be to Fabric v3.0

johannww committed:
Delete integration/ed25519 folder and revert vendor file

On previous commits, I made a changed on the vendored package
"vendor/github.com/hyperledger/fabric-config/configtx/signer.go",
since the tests on "integration/ed25519" required such change.

However, since this folder is being deleted and changes to the vendored
package would require another pull request, I decided to revert the
changes on the file.

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

johannww committed:
Update Channel capabilities and msp version to V3_0

As requested by the fabric Code Owner yacovm, the
capabilities and the msp version name must be renamed
to V3_0, because the ed25519-support will probably
enter as a feature in Fabric 3.0.

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

johannww committed:
Remove oldPeerRunner from integration test net

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

johannww committed:
Update ed25519 integration test case

Due to the renaming of the ed25519 Channel capabilities
to "v3_0" and other requests by fabric Code Owner yacovm,
the ed25519 integration test had to be updated.

Now the ed25519 integration test cases use a raft network with
three orderers.

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

johannww committed:
Dereference *ed25519.PublicKey and *ed25519.PrivateKey

This is an implementation correction. Since the methods
to store keys deals with interfaces, the ed25519 pointers
were not causing errors on compilation. However, ed25519
key pointers would cause errors in execution time.

Even though these methods to store keys are not used in
production, they might be one day.

Golang deals with ECDSA keys as pointers but not with
ED25519 and I cannot see a reason for that. However,
I believe that dealing with pointers, when it is possible,
is better for memory performance.

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

johannww committed:
Removing keyalg flag from cryptogen

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

johannww committed:
Handling ed25519 cast error

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

bestbeforetoday commented on cmd/common/signer/signer.go:
```diff
@@ -94,10 +95,18 @@ func validateEnrollmentCertificate(b []byte) error {
+	case *ecdsa.PrivateKey:
+		return signECDSA(si.key.(*ecdsa.PrivateKey), digest)
+	case ed25519.PrivateKey:
+		return ed25519.Sign(si.key.(ed25519.PrivateKey), digest), nil
```
Is this correct? The godoc suggests that, while [ecdsa sign takes a hash/digest](https://pkg.go.dev/crypto/ecdsa@go1.18.4#Sign), the [ed25519 sign requires the entire message bytes](https://pkg.go.dev/crypto/ed25519@go1.18.4#Sign). There is more detail in the PrivateKey Sign method doc:

https://pkg.go.dev/crypto/ed25519@go1.18.4#PrivateKey.Sign

yacovm replied:
based on the tests of the ed25519 package, you need to pass a message and not a hash. 

And this means that the verifier needs to be changed accordingly.

johannww replied:
I am evaluating a possible solution, as I believe these requirements will lead to changes in the Signer interface

https://github.com/johannww/fabric-1/blob/9fdc4f0c3ac92a7c95ab0099ac6535a1fc4410f0/bccsp/signer/signer.go#L76-L78

So functions like the one below can be called:

https://github.com/johannww/fabric-1/blob/9fdc4f0c3ac92a7c95ab0099ac6535a1fc4410f0/msp/identities.go#L255-L278

johannww replied:
I think it is fixed. Plus, I forced-pushed some commits because they were not signed off.

johannww committed:
Fix typo

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

johannww committed:
Update signer interface documentation and params

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

johannww committed:
Fix ed25519 signature

Previously, the golang ed25519 library was being used to
sign hashes. However, the golang ed25519 library must have
the full message as parameter and it calculates the hash
in the "ed25519.Sign()" function.

To solve that problem, the Sign and Verify methods in
"msp/identities.go" calculate the hash only if the
certificate does not have a ed25519 key.

Some parameters' names were also changed to keep the
coherence with the ed25519 Sign and Verify methods.

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

johannww committed:
Add identities ed25519 test cases

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

johannww committed:
Add ed25519 signer test case

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

johannww committed:
Fix fileks and keys for ed25519

Changes regarding dealing with pointers to
ed25519 keys versus the values. This is a
concern because golang generally deals
with ed25519 keys by value.

This causes a difference between the default
way to deal with ecdsa keys (pointers) and
ed25519 ones (values).

As I implemented, the ed25519 keys storing and parsing
follow the general rules:

1) Ed25519 keys are parsed from files as values (ed25519.PrivateKey or
ed25519.PublicKey) but are eventually converted to pointers
to fit the structs ed25519PrivateKey and ed25519PublicKey.

2) To store ed25519 keys in the fileks, they are dealt with as pointers
until the marshalling method, from the x509 golang library. This fits
test cases for storing "nil" keys, that would cause panic if the key
were treated as value.

Fileks test cases for ed25519 were added.

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

johannww committed:
Gen ed25519 or ecdsa crypto for users

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

johannww committed:
Gen ed25519 or ecdsa crypto for templates

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

bestbeforetoday commented on gossip/discovery/discovery_impl.go:
```diff
@@ -735,9 +735,9 @@ func (d *gossipDiscoveryImpl) expireDeadMembers(dead []common.PKIidType) {
-		d.comm.CloseConn(&member2Expire)
+	for i := range deadMembers2Expire {
+		d.logger.Warning("Closing connection to", deadMembers2Expire[i])
+		d.comm.CloseConn(&deadMembers2Expire[i])
```
I don't see the benefit of changing the code to look up the value twice by index instead of just directly using the value provided by for-range, as the code did before

yacovm replied:
I don't see how this is related to the pull request's intention.

johannww replied:
Do not consider that. This is the result of a forced push that I have done to sign my commits. I accidentally signed commits that were not mine. I will fix that and probably force-push again.

yacovm replied:
no worries

johannww committed:
Add PublicKeyAlgorithm to default spec

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

johannww committed:
Enhance warning for unsupported key

The warning in the logs when a key algorithm was not
supported was to big. Only the common names are used
now, instead of the whole subject and issuer.

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

yacovm commented on integration/lifecycle/capabilities_endorse_test.go:
```diff
@@ -0,0 +1,261 @@
+/*
```
Is there any specific handling for ed25519 that influences lifecycle specifically?

johannww replied:
No. I just put it in the lifecycle folder because it seemed the most related integration test "family", considering your previous  #suggestions:
> I think a **single** integration test that:
> 
> * Has peers and orderers in it
> * Transacts on a channel
> * The channel contains both ECDSA and ed25519 identities and they're both used in the endorsement policy
> 
> Is sufficient.

yacovm replied:
i see, ok

johannww committed:
Update capabilities test to ginkgo v2

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

johannww committed:
Fix checkstyle complaints

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

johannww committed:
Fix channelV30 test for MSPv3_0

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

johannww committed:
Fix cryptogen default config and test

Some tabs were introduced on previous commits
causing a parsing error on the defaultConfig.
A test was added to prevent this error from
happening again.

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

johannww commented:
One important issue:

I am implementing Ed25519 support for [fabric-gateway](https://github.com/johannww/fabric-gateway) using the code in this pull request as my Hyperledger Fabric. I created the crypto-material by setting field PublicKeyAlgorithm to **ed25519**, instead of **ecdsa** as here: 

https://github.com/johannww/fabric-gateway/blob/81f5ee44f37f7148500f1039e5b33bd4f17d8f4d/scenario/fixtures/crypto-material/crypto-config.yaml#L41-L45

This generated ed25519 TLS keys for the peers. When running the scenario-test, only the java-gateway failed in a scenario with ed25519 TLS keys, due to the library [netty](https://github.com/netty/netty/issues/10916) (see the issue). I see 2 possible ways to go from here, regarding changes to fabric and/or gateway: 

1. Disable ed25519 certificates for TLS explicitly in fabric and prevent cryptogen from generating ed25519 TLS certificates.
2. Ignore Java limitation and issue a warning when errors related to the grpc and netty happen. Count on netty developers to implement ed25519 support soon.
3. Any other suggestions?

bestbeforetoday commented:
> When running the scenario-test, only the java-gateway failed in a scenario with ed25519 TLS keys, due to the library [netty](https://github.com/netty/netty/issues/10916) (see the issue).

I noticed this JDK issue, which says TLS support for EdDSA signatures was added in Java 16:

https://bugs.openjdk.org/browse/JDK-8254709

Any chance that it worked with Java 17?

> I see 2 possible ways to go from here, regarding changes to fabric and/or gateway:
> 
> 1. Disable ed25519 certificates for TLS explicitly in fabric and prevent cryptogen from generating ed25519 TLS certificates.
> 2. Ignore Java limitation and issue a warning when errors related to the grpc and netty happen. Count on netty developers to implement ed25519 support soon.
> 3. Any other suggestions?

I think I would be happy to proceed with only Java clients not currently able to establish TLS connections to peers that use Ed25519 keys. There is precedent for certain crypto only being supported by certain client, such as Idemix. This is clearly a bug outside of our control and should get fixed at some point.

yacovm commented:
> One important issue:
> 
> I am implementing Ed25519 support for [fabric-gateway](https://github.com/johannww/fabric-gateway) using the code in this pull request as my Hyperledger Fabric. I created the crypto-material by setting field PublicKeyAlgorithm to **ed25519**, instead of **ecdsa** as here:
> 
> https://github.com/johannww/fabric-gateway/blob/81f5ee44f37f7148500f1039e5b33bd4f17d8f4d/scenario/fixtures/crypto-material/crypto-config.yaml#L41-L45
> 
> This generated ed25519 TLS keys for the peers. When running the scenario-test, only the java-gateway failed in a scenario with ed25519 TLS keys, due to the library [netty](https://github.com/netty/netty/issues/10916) (see the issue). I see 2 possible ways to go from here, regarding changes to fabric and/or gateway:
> 
>     1. Disable ed25519 certificates for TLS explicitly in fabric and prevent cryptogen from generating ed25519 TLS certificates.
> 
>     2. Ignore Java limitation and issue a warning when errors related to the grpc and netty happen. Count on netty developers to implement ed25519 support soon.
> 
>     3. Any other suggestions?

Fabric's TLS stack is generally up to the Go standard library, so if you use a Java client willingly, it is up to you to be able to connect correctly to the TLS stack of the nodes.

Generally, as a rule of thumb - Fabric core **sets the standard**, and the clients track the Fabric core, and not the other way around.

johannww committed:
Add CA's PublicKeyAlgorithm to default config

This commit adds the field PublicKeyAlgorithm to
cryptogen's default config. CAs crypto-material keys
might be generated using ed25519 or ecdsa.

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

johannww committed:
Add ecdsa as default PublicKeyAlgorithm for CA

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

yacovm commented:
@johannww what's the status of this? is this still work in progress?

johannww commented:
From my perspective it is ready.On 23 Dec 2022, at 08:16, yacovm ***@***.***> wrote:﻿
@johannww what's the status of this? is this still work in progress?

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

bestbeforetoday commented:
@johannww I'm not a maintainer of this repository so I don't know if there is any other reason this PR has not been reviewed, but it does show as having merge conflicts with the **main** branch that need to be resolved.

johannww committed:
Fix UpdateConfig() calls

Since the last merge of this working with the main
branch (ff4eb12df), the recent commit (14c3a0c92) introduced
new calls for the function nwo.UpdateConfig(). We fix these new calls
by passing a 'nil' ordererSigners argument.

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

denyeart commented:
Thanks for resolving the conflicts.

@yacovm Could you take another look at this one?

yacovm commented:
> Thanks for resolving the conflicts.
> 
> @yacovm Could you take another look at this one?

I want to merge this PR very much, as eddsa has additional benefits beyond having another signature algorithm - for example it has a much faster threshold signature algorithm than ECDSA. 

I'm a bit overloaded this week and the next, but I will try to find some time.

johannww commented:
For the fabric-gateway, I think it is better to wait for the release 3.0. It seems that a lot has changed (ubuntu docker images/fabric-tools does not have **go** binary anymore). The main ed25519 implementations are done, but I will wait for the gateway to catch up to Fabric 3.0 to offer the pull request.

bestbeforetoday commented:
> For the fabric-gateway, I think it is better to wait for the release 3.0. It seems that a lot has changed (ubuntu docker images/fabric-tools does not have **go** binary anymore). The main ed25519 implementations are done, but I will wait for the gateway to catch up to Fabric 3.0 to offer the pull request.

I don't really understand the motivation for waiting for Fabric v3, and why there was no desire to include this in the Fabric v2.5 LTS release. What is the blocker to both reviewing and delivering this PR?

denyeart commented:
Also, note that Go binary has recently been re-added to fabric-tools image in main branch and release-2.5 branches - https://github.com/hyperledger/fabric/pull/4176.

I agree that core Fabric support can lead SDK support.

mkhattat commented:
@johannww I tried to test your pull request with the recent release of fabric-gateway v1.3.0 (@bestbeforetoday ) which added support for Ed25519 signer implementations but I had no success. Here is what I did:

I used the `asset-transfer-basic` from `fabric-samples` repository to test. I mainly used the default settings. The only thing I changed was in the file `fabric-samples/test-network/organizations/cryptogen/crypto-config-org1.yaml`:

```yaml
PeerOrgs:
  - Name: Org1
    Domain: org1.example.com
    EnableNodeOUs: true
    Template:
      Count: 1
      SANS:
        - localhost
    Users:
      Count: 1
      PublicKeyAlgorithm: ed25519

```

I used the following commands (inside the `test-network` folder) to create a channel and deploy it:
```bash
export FABRIC_CFG_PATH=$PWD/../config/
export CORE_PEER_TLS_ENABLED=true
export CORE_PEER_LOCALMSPID="Org1MSP"
export CORE_PEER_TLS_ROOTCERT_FILE=${PWD}/organizations/peerOrganizations/org1.example.com/peers/peer0.org1.example.com/tls/ca.crt
export CORE_PEER_MSPCONFIGPATH=${PWD}/organizations/peerOrganizations/org1.example.com/users/User1@org1.example.com/msp
export CORE_PEER_ADDRESS=localhost:7051
export CORE_VM_DOCKER_ATTACHSTDOUT=true

./network.sh down
./network.sh up createChannel -c mychannel
./network.sh deployCC -ccn basic -ccp ../asset-transfer-basic/chaincode-go -ccl go
```

After deploying the channel, I compiled and ran the `assetTransfer` application from `fabric-samples/asset-transfer-basic/application-gateway-go/assetTransfer.go` folder. In the `initLedger` step it fails with the following error:

```
failed to submit transaction: rpc error: code = Aborted desc = failed to endorse transaction, see attached details for more info.

check if creator is a valid certificate - access denied: channel [mychannel] creator org [Org1MSP]: could not validate identity's public key algorithm: Ed25519 is not supported.
```

Could you please help me out here if I am missing some configuration?

johannww commented:
@mkhattat, I don't have much experience with the test network, but you should look [here in the config.tx file](https://github.com/hyperledger/fabric-samples/blob/7671bdd7c67889560bee1dd20d40aa30f1d7157e/test-network/configtx/configtx.yaml#L120). You should replace "V2_0" for "V3_0", because ed25519 is, **for now**, a fabric 3.0 feature. This might change, as discussed here in this pull request.

bestbeforetoday commented:
@mkhattat Did you have any luck getting the asset-transfer-basic sample to with with the Fabric configuration change suggested by @johannww above?

One other thing to note is the **hash** expectations of different signature algorithms, documented [here](https://pkg.go.dev/github.com/hyperledger/fabric-gateway@v1.4.0/pkg/identity#NewPrivateKeySign).

> Note that the Sign implementations have different expectations on the input data supplied to them.
>
> The ECDSA signers operate on a pre-computed message digest, and should be combined with an appropriate hash algorithm. P-256 is typically used with a SHA-256 hash, and P-384 is typically used with a SHA-384 hash.
>
> The Ed25519 signer operates on the full message content, and should be combined with a NONE (or no-op) hash implementation to ensure the complete message is passed to the signer.

By default the client API applies a SHA-256 hash to the message to create a message digest, and that message digest is passed to the signer. This is the desired behaviour for an ECDSA signer, which operates on a pre-computed digest. The Ed25519 signer computes the digest as part of the signing operation so needs to receive the complete message bytes.

To use an Ed25519 signer, it is not sufficient to only provide an Ed25519 signing implementation as an option to [Connect](https://pkg.go.dev/github.com/hyperledger/fabric-gateway/pkg/client#Connect). You must also provide the [WithHash](https://pkg.go.dev/github.com/hyperledger/fabric-gateway/pkg/client#WithHash) option with a [NONE](https://pkg.go.dev/github.com/hyperledger/fabric-gateway/pkg/hash#NONE) hash implementation. This causes the unaltered message bytes to be passed to the Ed25519 signing implementation instead of a pre-computed digest.

bestbeforetoday commented on cmd/common/signer/signer.go:
```diff
@@ -94,10 +95,18 @@ func validateEnrollmentCertificate(b []byte) error {
 
 func (si *Signer) Sign(msg []byte) ([]byte, error) {
 	digest := util.ComputeSHA256(msg)
```
The digest is always being computed, even though it is only used by one of the switch cases below. It should be moved into that switch case to avoid redundant work

johannww committed:
compute SHA256 only for ECDSA keys

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

johannww committed:
fix comments assuming ECDSA-only support

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

bestbeforetoday commented:
@hyperledger/fabric-core-maintainers Now that Fabric v3 is the next scheduled release from the main branch codebase, what is the blocker to this PR being reviewed and merged? This submission has been open for approaching 2 years so it would be good to make some kind of decision on what is to be done with it.

johannww commented:
Complementing the @bestbeforetoday  request, let me clarify the importance of this pull request for me and Hyperledger Fabric. I implemented the code in this pull request for a Brazilian Government project I work for. Brazil's public-key infrastructure [does not allow NIST curves anymore](https://repositorio.iti.gov.br/instrucoes-normativas/IN2022_22_DOC-ICP-01.01.htm). The first option was to adopt the Brainpool curves, but Go did not implement it. Thus, we opted to implement ed25519 support and offer this pull request.

Generally, I believe that supporting ed25519 can expand Hyperledger Fabric's adoption for similar cases. I guess that there are many other valid incentives. For instance, the Brazilian CBDC [pilot](https://github.com/bacen/pilotord-kit-onboarding/blob/main/arquitetura.md) is using Hyperledger Besu. Consider that, as I believe it is in your and the community's best interest to evaluate this pull request.

yacovm commented on discovery/cmd/stub_test.go:
```diff
@@ -45,6 +45,31 @@ func TestClientStub(t *testing.T) {
+		},
+		TLSConfig: comm.Config{},
+	}, req)
+	require.Contains(t, err.Error(), "Unimplemented desc = unknown service discovery.Discovery")
```
but why would we want to test this case? I think it's not a very interesting case to demonstrate compatibility with ed25519. 

The discovery code base has nothing to do with the signature algorithm. I don't think we should be expanding these unit tests for this PR.

johannww replied:
I was lost as to what specifically test, so I probably implemented it due to the presence of a _SignerConfig_ loading ECDSA key/certificate. I will remove it.

yacovm commented on common/capabilities/channel.go:
```diff
@@ -86,7 +86,9 @@ func (cp *ChannelProvider) HasCapability(capability string) bool {
 func (cp *ChannelProvider) MSPVersion() msp.MSPVersion {
 	switch {
-	case cp.v143 || cp.v20 || cp.v30:
+	case cp.v30:
```
we need an integration test that:

(1) Spins up a v2 channel level network, and transacts with ECDSA. 
(2) Upgrades channel level to v3 and then transacts with ed25519.
(3) Joins a new peer to the channel without a snapshot, and make sure the peer's state is the same as the rest of the peers in the network.

johannww replied:
I think this [test](https://github.com/johannww/fabric-1/blob/2a82a06294273bb16a27b42d195fdf04d5e79a0e/integration/lifecycle/capabilities_endorse_test.go#L80-L288) is ready now.

yacovm commented on internal/cryptogen/csp/csp.go:
```diff
@@ -64,17 +65,29 @@ func parsePrivateKeyPEM(rawKey []byte) (*ecdsa.PrivateKey, error) {
+	_, isEcdsa := key.(*ecdsa.PrivateKey)
+	_, isEd25519 := key.(ed25519.PrivateKey)
+	if !isEcdsa && !isEd25519 {
+		return nil, errors.New("pem bytes do not contain an EC nor ED private key")
```
EC refers to an Elliptic Curve, not ECDSA. Just keep it EC.

yacovm replied:
(or just write ECDSA / ed25519)

yacovm commented on integration/lifecycle/capabilities_endorse_test.go:
```diff
@@ -0,0 +1,261 @@
+	})
+})
+
+func changePeerOrOrdererToEd25519(network *nwo.Network, orgName, entityname string) {
```
why do you need to read the entire config? Can't you just shut down the node, and replace its MSP folder and start it again?

johannww replied:
The crypto material on integration tests are created by cryptogen, from the template here: [nwo crypto template](https://github.com/hyperledger/fabric/blob/5cb00bef7c4c499481b5538034a1975dbe3c9319/integration/nwo/template/crypto_template.go#L9-L11). To avoid major changes on the default network definitions, I use the default [MultiNodeEtcdRaft](https://github.com/hyperledger/fabric/blob/5cb00bef7c4c499481b5538034a1975dbe3c9319/integration/nwo/standard_networks.go#L234), but add another peer to have a ed25519 key. Also, I provide an ed25519 key for **orderer3**. Both are enabled by modifying the cryptogen config for the extra peer and **orderer3**, which is the action performed by this function. 

This prevents me from creating another cryptogen template for an ed25519 integration test, or adding another network definition.

Do you think there is a better way?

johannww replied:
By the way, I am fixing this test. I designed it when there was a system channel.

yacovm replied:
If all you want is to add a peer with a different key and certificate, you can generate the key and certificate and then have the CA sign the key beforehand: https://github.com/hyperledger/fabric/blob/main/integration/raft/cft_test.go#L1167-L1200

yacovm commented on integration/lifecycle/capabilities_endorse_test.go:
```diff
@@ -0,0 +1,261 @@
+		)
+
+		By("invoking the chaincode again, but expecting a failure")
+		RunInvokeAndExpectFailure(network, orderer, "My_1st-Chaincode", "(ENDORSEMENT_POLICY_FAILURE)", endorsers...)
```
can we be more specific about the failure? Is there any signature specific failure in the log of the node that we can capture here?

johannww replied:
The whole return message is: "Error: transaction invalidated with status (ENDORSEMENT_POLICY_FAILURE) - proposal response: <nil>". The test is meant to demonstrate that an ed25519 cannot endorse transactions in a channel with V2_0 capabilities.

yacovm commented:
Thanks for not giving up on this PR. I did a superficial pass and gave some comments. 

Please rebase on top of latest main and let's also have @ale-linux and @adecaro  and @denyeart  take a look at this.

yacovm commented:
By the way, what about TLS? Have you tried spinning up a node with TLS certificates containing ed25519 keys? I am sure it should be possible as we don't do anything special in our TLS stack regarding keys, but I would like to ask that regardless.

bestbeforetoday commented:
> By the way, what about TLS? Have you tried spinning up a node with TLS certificates containing ed25519 keys? I am sure it should be possible as we don't do anything special in our TLS stack regarding keys, but I would like to ask that regardless.

Testing with Ed25519 TLS was mentioned in [this comment](https://github.com/hyperledger/fabric/pull/3343#issuecomment-1327967248) above. The only issue mentioned was with Java client applications due to a bug / limitation in the Netty library used by gRPC-Java.

yacovm commented:
> > By the way, what about TLS? Have you tried spinning up a node with TLS certificates containing ed25519 keys? I am sure it should be possible as we don't do anything special in our TLS stack regarding keys, but I would like to ask that regardless.
> 
> Testing with Ed25519 TLS was mentioned in [this comment](https://github.com/hyperledger/fabric/pull/3343#issuecomment-1327967248) above. The only issue mentioned was with Java client applications due to a bug / limitation in the Netty library used by gRPC-Java.

oh I totally forgot about it, I even replied to the comment :-) 

I still concur with Yacov from the past:

> Generally, as a rule of thumb - Fabric core sets the standard, and the clients track the Fabric core, and not the other way around.

johannww committed:
fix key type error message

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

johannww committed:
remove an unnecessary test

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

johannww committed:
replace ioutil with os

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

johannww committed:
Improve ed25519 lifecycle integration test

    - instead of changing cryptogen config, we
    give ed25519 keys by changing the certificate
    and the keys.
    - Now, the test is compatible with a network without
    a system channel.
    - Other improvements were made

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

johannww committed:
refactor

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

johannww committed:
remove unecessary line from ed25519 test

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

johannww committed:
fix nwo.UpdateConfig() call from discovery test

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

johannww committed:
use gomega funcs instead of println

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

bestbeforetoday commented:
@johannww likely because other changes have been merged in the meantime but this PR is showing as having merge conflicts that need to be resolved. You might need to rebase your change on the main branch and force-push an updated commit. Hopefully then @yacovm can review.

yacovm commented:
I can review but I think it's also good if @adecaro and @ale-linux will both take a look :-)

johannww commented:
@bestbeforetoday I guess I will have to open a pull request in fabric-lib-go also. The deleted bccsp folder seems to be the only problem.

johannww commented:
I rebased on main branch again. Notice that my **bccsp** modifications were moved to the [fabric-lib-go vendor folder](https://github.com/johannww/fabric-1/tree/ed25519-support/vendor/github.com/hyperledger/fabric-lib-go) but the changes were not accepted at fabric-lib-go repo yet. A [pull request](https://github.com/hyperledger/fabric-lib-go/pull/29) was opened there.

bestbeforetoday commented:
@hyperledger/fabric-core-maintainers Please can somebody review hyperledger/fabric-lib-go#29 so this PR can be completed?

yacovm commented:
> @hyperledger/fabric-core-maintainers Please can somebody review [hyperledger/fabric-lib-go#29](https://github.com/hyperledger/fabric-lib-go/pull/29) so this PR can be completed?

I reviewed but i would like @adecaro to also take a look at it

bestbeforetoday commented:
@johannww Now that hyperledger/fabric-lib-go#29 has been been merged, I guess this PR needs updating to pull in the new fabric-lib-go dependency version?

johannww committed:
bump fabric-lib-go to v1.1.3

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

bestbeforetoday commented:
@johannww I have no idea why, but your last commit does not seem to have successfully triggered a build. I do see some other pull request builds triggered since then so perhaps whatever the problem was is now resolved. You should be able to trigger a build by just amending (`git commit --amend`) and re-pushing your last commit.

bestbeforetoday commented:
@yacovm @denyeart @adecaro @ale-linux this PR is now building successfully against the BCCSP updates delivered in hyperledger/fabric-lib-go#29. Can you re-review?

yacovm commented on integration/nwo/cryptogen_config.go:
```diff
@@ -0,0 +1,58 @@
+	Users         UsersSpec    `yaml:"Users"`
+}
+
+type CryptogenConfig struct {
```
where is this used?

johannww replied:
On previous tests, I used cryptogen to give ed25519 credentials. It has no use now. It will be removed.

yacovm commented on integration/nwo/deploy.go:
```diff
@@ -499,7 +499,29 @@ func EnableCapabilities(network *Network, channel, capabilitiesGroup, capabiliti
+		Value: protoutil.MarshalOrPanic(
+			&common.Capabilities{
+				Capabilities: map[string]*common.Capability{
+					capabilitiesVersion: {},
```
I'm confused, where do you enable the V3_0 capability here?

johannww replied:
I followed the example of [./integration/e2e/e2e_test.go#L197-L198](https://github.com/hyperledger/fabric/blob/05b604cc64eeed2684e917a7d34730d876a2ee59/integration/e2e/e2e_test.go#L197-L198), which is a pattern repeated in many integration tests.  V3_0 capabilities are enabled by line **518**.

yacovm replied:
oh I see, the capabilitiesVersion you pass is set to V3_0

yacovm commented on integration/lifecycle/capabilities_endorse_test.go:
```diff
@@ -0,0 +1,346 @@
+	})
+
+	It("invoke chaincode after upgrading Channel to V3_0 and add ed25519 peer and orderer", func() {
+		networkConfig := nwo.MultiNodeEtcdRaft()
```
So we have a single organization with 2 orderers of ECDSA and 1 orderer that is offline, and later is brought online to be ed25519 ?

johannww replied:
Almost that. Initially, we have 3 ECDSA orderers from a single organization. Later, orderer3 is killed, given an ed25519 key, and restarted. In contrast, there are 2 ECDSA peers from different orgs and an ed25519 offline one. The ed25519 peer is only started once, directly with ed25519 keys.

yacovm commented on msp/identities_test.go:
```diff
@@ -0,0 +1,215 @@
+// sign the DIGEST
+// The same applies to the Verify function
+func TestSignatureAlgorithms(t *testing.T) {
+	gt := gomega.NewGomegaWithT(t)
```
not that it matters, but, any reason you use gomega here and not "github.com/stretchr/testify/require" ? 

The entire package is using the latter

johannww replied:
I did not notice and was probably inspired by a gomega test file. I can change that to keep coherence.

denyeart commented:
@johannww I just wanted to say thank you for sticking with this pull request. I acknowledge that it has taken a  very long time to fully review it, but this is indeed on our list for the upcoming v3.0 release.

Do you have any interest in adding ed25519 support to fabric-ca for certificate issuance?

johannww committed:
remove uneeded file

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

johannww committed:
fix integration test chaincode deploy

Signed-off-by: Johann Westphall <johannwestphall@gmail.com>

johannww commented:
@denyeart You're welcome. I remember. I tried to add ed25519 to the fabric-ca previously but stumbled into a problem with cloudfare cfssl support for ed25519: [https://github.com/cloudflare/cfssl/pull/1097#issuecomment-1332834508](https://github.com/cloudflare/cfssl/pull/1097#issuecomment-1332834508).

Here is my old fork of fabric-ca: [https://github.com/johannww/fabric-ca](https://github.com/johannww/fabric-ca)

This seems to be solved and merged. I should take a look into it.

yacovm commented:
ok it looks good to me, let's have @ale-linux or @adecaro also review it.

bestbeforetoday commented:
> ok it looks good to me, let's have @ale-linux or @adecaro also review it.

How long do we wait for a review before merging?

denyeart commented:
> > ok it looks good to me, let's have @ale-linux or @adecaro also review it.
> 
> How long do we wait for a review before merging?

I talked to @adecaro the other day and he confirmed he will take a look so that we can get this in for v3.0... I do think it is valuable to have one of the original MSP authors take a look.

adecaro commented:
LGTM 👍

denyeart reviewed (approved):
@johannww Thanks again for the contribution and apologies for the delays, this is now ready for merge for inclusion in Fabric v3.0!

johannww commented:
@denyeart you're welcome! Thank you also for reviewing big pull request.

@bestbeforetoday thank you for ensuring that this pull request remained active throughout this period.

@yacovm thank you for all the reviews and hints given.
